### ISSUE TITLE
Make fails on Ubuntu 18.04

### ISSUE THREAD

**Issue by gharris999** [jqlang/jq#1659]
`./.libs/libjq.a(builtin.o): In function `f_pow10':
/home/daadmin/StdUtils/jq/src/libm.h:253: undefined reference to `pow10'
collect2: error: ld returned 1 exit status
Makefile:984: recipe for target 'jq' failed
make[2]: *** [jq] Error 1
make[2]: Leaving directory '/home/daadmin/StdUtils/jq'
Makefile:1148: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/daadmin/StdUtils/jq'
Makefile:825: recipe for target 'all' failed
make: *** [all] Error 2
`
Is this the same issue as #1565 ??

**Comment by gharris999** [comment]
Output of..
./configure --with-oniguruma=builtin --disable-maintainer-mode
..at
https://pastebin.com/irSA4qVS

**Comment by doronbehar** [comment]
I'm experiencing a very similar problem here on Arch Linux,

Here is my trace:
```
make[2]: Entering directory '/var/cache/aursync/jq-git/src/jq'
  CCLD     jq
./.libs/libjq.a(util.o): In function `get_home':
util.c:(.text+0x3f2): warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
./.libs/libjq.a(builtin.o): In function `f_pow10':
builtin.c:(.text+0x50c6): undefined reference to `pow10'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:984: jq] Error 1
make[2]: Leaving directory '/var/cache/aursync/jq-git/src/jq'
make[1]: *** [Makefile:1147: all-recursive] Error 1
make[1]: Leaving directory '/var/cache/aursync/jq-git/src/jq'
make: *** [Makefile:824: all] Error 2
```

Here is the [output of `./configure --prefix=/usr --with-oniguruma=builtin --disable-maintainer-mode`](https://gist.github.com/doronbehar/bda71f564f16b506c10897c10baad13b)

**Comment by marcelpaulo** [comment]
Also happening with a fresh install of Xubuntu 18.04. 

The output of `./configure --with-oniguruma=builtin` has:

```
checking for pow10... yes
checking for pow... yes
```

But `make` fails:

```
./.libs/libjq.a(builtin.o): In function `f_pow10':
/home/paulo/src/jq/src/libm.h:253: undefined reference to `pow10'
collect2: error: ld returned 1 exit status
```

**Comment by marcelpaulo** [comment]
I think I discovered why this error occurs, although I still don't know how to fix it. I ran `remake -x` in place of `make`, to see which command fails, it's:

```
libtool --silent --tag=CC   --mode=link gcc -Wextra -Wall -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-function    -I./modules/oniguruma/src -g -O2 -static-libtool-libs   -o jq src/main.o libjq.la -lm
```

It includes the math library `-lm`, so it should find **pow10**, but … the [man page](http://man7.org/linux/man-pages/man3/pow10.3.html) for pow10 says:

> These functions first appeared in glibc in version 2.1.  Since glibc 2.27, the use of these functions in new programs is no longer supported.

The Ubuntu package which contains the math library is **libc6**. Looking up libc6 in the [Ubuntu packages archive](https://packages.ubuntu.com/search?keywords=libc6), I can see that Ubuntu 17.10 has version **2.26-0ubuntu2.1**, while Ubuntu 18.04 has **2.27-3ubuntu1**, so pow10 should not be included. Confirming, pow10 isn't defined in any **libc6-dev** header file:

```
+paulo@monk:~/src/jq$ grep -R '\<pow10\>' /usr/include
+paulo@monk:~/src/jq$ 
```

So, it seems that `configure` mistakenly detects a non-existent pow10:

```
checking for pow10... yes
```

As a workaround, we should be able to run `configure` declaring beforehand that pow10 doesn't exist. I don't know if that's possible or how to do it. I'll keep on digging and post back any promising results.

**Comment by marcelpaulo** [comment]
A crude hack to workaround this issue:

1. Remove the line in **_configure.ac_** that checks for pow10:
    `AC_CHECK_MATH_FUNC(pow10,[.5])`
1. Run again:
    ```
    autoreconf -fi
    ./configure --with-oniguruma=builtin
    ```
1. Clean the previous compilation output:
    `make clean`
1. Continue with the build, that is:
    ```
    make -j8
    make check
    sudo make install
    ```

**Comment by gharris999** [comment]
With that hack, make succeeds for me.  However, on my system, the jq.test, called by make check, seems to enter into an infinite loop, pegging the CPU at 100%.  On the same hardware, running ubuntu server 16.04 and jq built with the same configuration on that system, jq.test completes in about 58 seconds.

**Comment by gharris999** [comment]
Actually, let me amend that:  the jq.test did complete, but it took 20' 53" to complete.

So, at least on my system, removing pow10 from the build significantly slows down some operations.

**Comment by marcelpaulo** [comment]
> However, on my system, the jq.test, called by make check, seems to enter into an infinite loop, pegging the CPU at 100%

Puzzling. I suppose that ignoring pow10 means only that it was replaced by **pow**, which shouldn't account for this.

Digging a little bit more, I discovered that macro `AC_CHECK_MATH_FUNC` is defined using macro `AC_FIND_FUNC_NO_LIBS`, which in turn is defined using macro `AC_FIND_FUNC_NO_LIBS2`, which then actually checks if a math function is present in a library. I couldn't decypher `AC_FIND_FUNC_NO_LIBS2`, but I could see that `AC_CHECK_MATH_FUNC` is defined like this:

```
AC_FIND_FUNC_NO_LIBS([$1], [m], [#include <math.h>], [$2])
```

which would lead one to believe that, with the definition `AC_CHECK_MATH_FUNC(pow10,[.5])` in *configure.ac*, it would check header ***math.h*** and library ***libm***. As I showed earlier, there's no declaration of **pow10** in any header files under _/usr/include_. So somehow this macro is failing.

The source file with the failing macro is: _config/m4/find-func-no-libs.m4_

**Comment by marcelpaulo** [comment]
> So, at least on my system, removing pow10 from the build significantly slows down some operations.

But the fact is: pow10 does seem to be available anymore since glibc 2.27. So it will have to be removed, either properly — detecting that it's not there —, or with a crude hack — not trying to detect it.

**Comment by gharris999** [comment]
Well, according to:

http://man7.org/linux/man-pages/man3/pow10.3.html

>  These functions return the value of 10 raised to the power x.

       Note well: These functions perform exactly the same task as the
       functions described in exp10(3), with the difference that the latter
       functions are now standardized in TS 18661-4:2015.  Those latter
       functions should be used in preference to the functions described in
       this page.

       These functions first appeared in glibc in version 2.1.  Since glibc
       2.27, the use of these functions in new programs is no longer
       supported.

So presumably the code could be fixed by using exp10 in place of pow10.  But for the life of me I can't figure out where pow10 gets called.

**Comment by marcelpaulo** [comment]
> So presumably the code could be fixed by using exp10 in place of pow10.

It makes sense.

But ... inspecting the symbols in the resulting jq binary:

```
+paulo@monk:~/src/jq$ nm -D jq | grep -E 'exp|pow'
                 U exp
                 U exp10
                 U exp2
                 U expm1
                 U frexp
                 U ldexp
                 U pow
```

**exp10** is being linked. I wouldn't know if the calls to pow10 are being replaced by pow or exp10.

**Comment by gharris999** [comment]
I suspect that pow10 is not being replaced by exp10.  I think exp10 is being called in another context.  There are references to exp10 in configure.ac and src/libm.h.

In any event, while the jq "torture test" may lag, jq compiled and linked with your hack excluding pow10 works perfectly for the simple purposes to which I put it.

Other than this little make glitch, the top entry in my jq wish list would be to have the parser ignore JSON comments.  It's kind of a pita to have to remember to always include jsonlint in my scripts to pre-process JSON data before feeding it to jq.

**Comment by marcelpaulo** [comment]
> jq compiled and linked with your hack excluding pow10 works perfectly for the simple purposes to which I put it

I'm almost ashamed to tell what I use jq for: to parse JSON output from [guessit](https://github.com/guessit-io/guessit), in a 1-screenful bash script that renames srt files to match their corresponding video files.

I'm happy that we managed to build it. It pains to upgrade OS and have stuff breaking.

**Comment by alex-ozdemir** [comment]
Wow, what a bug. We got bitten by GCC optimizing away function calls, even when the function being optimized away didn't actually exist. Anyways, the above PR fixes it.

**Comment by itchyny** [comment]
This issue has been resolved by #1702 (0c845aa291230cace495d06ac5465bd3ccb99319) in 2018, so closing.
